### ISSUE TITLE
Add quantified superclass to Bifunctor

### ISSUE THREAD

**Issue by Icelandjack** [haskell/core-libraries-committee#91]
Add a quantified constraint `forall a. Functor (bi a)` to connect `Bifunctor` to `Functor`. The superclass logically follows from the meaning of `Bifunctor` and Edward Kmett has done the same for `Profunctor` in his own library.

```haskell
{-# Language QuantifiedConstraints #-}

class (forall a. Functor (bi a)) => Bifunctor bi where
  ..
```

Here is a [comment from Kmett](https://www.reddit.com/r/haskell/comments/xjg6dz/superclasses_for_eq1_eq2_and_relaxed_instances/ip8mip7/) about how `Profunctor` is waiting on this change in `Bifunctor`:

> Once we have the superclass for `Bifunctor` we can fix the `Cayley`/`Tannen` split, which can then become one type.
>
> I stalled a bit in actually doing this on my own primarily because `Bifunctor` moved into base and outside of my ability to unilaterally make the change. (Which means the bifunctors package is currently stuck because I can't line it up with the profunctors package HEAD until that superclass is in place.)

Related issues
* #92 Add `Profunctor` with a quantified superclass
* #93 Add quantified superclass to `Bifoldable` and `Bitraversable`

**Comment by Bodigrim** [comment]
@Icelandjack I do not foresee technical objections, but are you up to an [impact assessment](https://github.com/haskell/core-libraries-committee/pull/89/files)?

**Comment by Icelandjack** [comment]
I made a merge request: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9060. I haven't used clc-stackage, will have to look into it in the morning.

The instances in `Data.Bifunctor` necessitated three `Functor` instances

```haskell
instance Functor ((,,,,) a b c d)
instance Functor ((,,,,,) a b c d e)
instance Functor ((,,,,,,) a b c d e f)
```

I took the liberty to making second a default definition (`second = fmap`) and weakening the minimal set from `{-# Minimal bimap | first, second #-}` to `{-# Minimal bimap | first #-}`.

**Comment by Icelandjack** [comment]
@Bodigrim Can you give me pointers for how to get *clc-stackage* running?

**Comment by Icelandjack** [comment]
Actually I probably don't have enough space on my computer to run that :smile:

**Comment by treeowl** [comment]
I have a significant concern about this proposal, even though I think it's ultimately the right thing: mismatched laziness for tuples. You write

```haskell
A partially applied 'Bifunctor' must be a 'Functor' and
-- the 'second' method must agree with that 'fmap'. From this it
-- follows that:
--
-- @'second' 'id' = 'id'@
```

Putting this more concisely, there should be a law

```haskell
@'second' = 'fmap'@
```

But that is not the case for tuples! We have

```haskell
fmap id (_|_ :: (a, b)) = _|_
second id (_|_ :: (a, b)) = (_|_, _|_)
```

The reason for this, as I understand it, is that the `bifunctors` package wanted `Data.Bifunctor.first` and `Data.Bifunctor.second` to agree with `Control.Arrow.first` and `Control.Arrow.second`, which are lazy to work well with `ArrowLoop`. But it's really quite a nasty mismatch, and much more so once `Functor` and `Bifunctor` are more formally tied together.

I would want to redefine the `Bifunctor` instances for tuples to be *strict* (which will also require making their `Biapplicative` instances strict again; I recently moved to have those changed for consistency with `Bifunctor`). Then I'd want a newtype like

```haskell
newtype LazyTup a b = LazyTup (a, b)
instance Functor (LazyTup a) where
  fmap f (LazyTup ~(a, b)) = (a, f b)
instance Bifunctor LazyTup where
  bimap f g (LazyTup ~(a, b)) = (f a, g b)
instance Biapplicative LazyTup where
  bipure a b = LazyTup (a, b)
  biliftA2 f g (LazyTup ~(a, b)) (LazyTup ~(c, d))= LazyTup ((f a c), (g b d))
```

**Comment by Bodigrim** [comment]
> Actually I probably don't have enough space on my computer to run that 😄 does submitting a CLC proposal really require people to compile all of Stackage which supposedly takes around 2 days?

Yes.

**Comment by treeowl** [comment]
@Bodigrim , that's pretty classist. Change your policy so people who aren't rich can still make proposals.

**Comment by Bodigrim** [comment]
@treeowl I'm not in a good humour today. If people expect unpaid volunteers to take their proposals seriously, they can certainly spare 48 machine hours and couple gigabytes of hard drive to prepare an impact assessment.

**Comment by Bodigrim** [comment]
> I have a significant concern about this proposal, even though I think it's ultimately the right thing: mismatched laziness for tuples.

Potentially we can split the proposal into two: one adding a superclass, and another adding a law `second = fmap` and fixing mismatch for tuples.

**Comment by treeowl** [comment]
@Bodigrim , as you've effectively banned me from making proposals from two different directions, that's not a viable path at all.

**Comment by Bodigrim** [comment]
@treeowl I'm sorry?

**Comment by L-as** [comment]
It seems like this is something that could be automated to some extent? Iceland Jack only has a toaster available AFAIK.

**Comment by treeowl** [comment]
I also have just one old laptop and I need to use it to compile stuff for work. I don't have the computational cycles or likely even storage available to compile all of Stackage.

**Comment by Bodigrim** [comment]
> It seems like this is something that could be automated to some extent?

It's kinda already automated to cloning `clc-stackage` and running `cabal build`, right? 

>  Iceland Jack only has a toaster available AFAIK.

If this toaster is capable to build GHC, it is likely capable to build Stackage as well. 

> I also have just one old laptop and I need to use it to compile stuff for work. I don't have the computational cycles or likely even storage available to compile all of Stackage.

I also have a single underpowered laptop with a very limited storage. It serves me well for CLC impact assessment.

**Comment by L-as** [comment]
With automated, I mean way by which you can easily submit a patch, and a remote server provisioned by e.g. the Haskell Foundation so that people with toasters or from poor or undeveloped countries can still contribute to Haskell. This is usually just CI, and compiling all of Stackage on PRs to base seems reasonable to me.

**Comment by chreekat** [comment]
> With automated, I mean way by which you can easily submit a patch, and a remote server provisioned by e.g. the Haskell Foundation so that people with toasters or from poor or undeveloped countries can still contribute to Haskell. This is usually just CI, and compiling all of Stackage on PRs to base seems reasonable to me.

I think we're 99% of the way there. GHC CI already compiles HEAD against head.hackage nearly every night, and head.hackage is nearly Stackage.

Adding a manual job to the pipeline that clones clc-stackage and builds it with the compiler built in an earlier step, a la the [perf-nofib job](https://gitlab.haskell.org/ghc/ghc/-/blob/master/.gitlab-ci.yml#L659), may be all that is required. At least for changes to boot libraries. *(Some details and opinions elided. Open an issue on the GHC tracker and/or a MR adding this job if you want to discuss it more. As much as I would like to just do it myself, I would also like to make CI overall more reliable, so I'm focusing on that first.)*

**Comment by Ericson2314** [comment]
Yeah I agree this feels very clearly like a thing where the Haskell Foundation should step in and make it as easy as possible. I think it is good if people feel it is easy to contribute to important library's design / it being harder to do regression tests doesn't feel like the right way to deal with Eternal September problems (which can be real but not saying anyone is trying on purpose to put up barriers entry).

**Comment by Bodigrim** [comment]
Proposers are welcome to do impact assessment themselves, ask a friend, pray for a miracle or HF. Anything goes. I don't buy the claim that it is disproportionately difficult; plenty of people figured out how to do it before.

> With automated, I mean way by which you can easily submit a patch, and a remote server provisioned by e.g. the Haskell Foundation so that people with toasters or from poor or undeveloped countries can still contribute to Haskell. This is usually just CI, and compiling all of Stackage on PRs to base seems reasonable to me.

Impact assessment is not a binary characteristic. If a package fails to build, a proposer is supposed to clone it with `cabal unpack`, patch locally, link into `cabal.project` and give it another go. This is an interactive process; CI approach does not really fit well here.

Anyway, this is all off topic. If there is an interest to discuss this digression further, please create a separate issue.

**Comment by sjakobi** [comment]
On the topic of building `clc-stackage`, note that you don't have to build all of it in one sitting. You can always Ctrl-C and continue at another time. Nor does the build process need to use all your cores, you can limit it to e.g. 2 processes with `-j2`.

It would be good to document how much disk space it requires though.

**Comment by Bodigrim** [comment]
Thanks @sjakobi, excellent points. I've updated https://github.com/Bodigrim/clc-stackage#how-to. 

@Icelandjack what do you think?

**Comment by Bodigrim** [comment]
@Icelandjack just a gentle reminder about your proposal here.

**Comment by Topsii** [comment]
### Issues Building clc-stackage
I tried to build `clc-stackage`, but I hit an error that I don't understand.
```
Tobias@DESKTOP-RODILQ3 MINGW64 /e/ghc-bifunctors/clc-stackage
$ cabal build -w "E:\ghc-bifunctors\ghc\_build\stage1\bin\ghc-bifunctor.exe" --keep-going --constraint="HsOpenSSL +use-pkg-config" --constraint="pcre-light +use-pkg-config" --constraint="hmatrix +openblas" --constraint="hexpat +bundle"
Build profile: -w ghc-9.2.4 -O1
In order, the following will be built (use -v for more details):
 - clc-stackage-0.1.0.0 (lib) (first run)
cabal-3.6.2.0.exe: Failed to build clc-stackage-0.1.0.0. The failure occurred
during the configure step. The exception was:
C:\ghcup\bin\cabal-3.6.2.0.exe: createProcess: does not exist (No such file or
directory)
```
Though I think I have built most/all packages. [Here's the `-v3` output.](https://github.com/haskell/core-libraries-committee/files/9781021/out14.txt)

I have removed 70 packages from `build-depends`. Most are only supposed to work on Unix. A few have `build-type: configure` scripts that I could not get to work or dependencies that I was not able to supply. The `katip` packages and `beam-sqlite` only work with an older version of the `Win32` package. `fsevents` is Mac only. `ki`, `path-extra` and `krank` has template haskell failures that seemed unrelated:
<details>
  <summary>Click to see TH failures</summary>

```
    • Exception when trying to run compile-time code:
        InvalidAbsDir "/"
CallStack (from HasCallStack):
  error, called at src\\Path\\Include.hs:759:20 in path-0.9.2-inplace:Path.Windows
      Code: template-haskell-2.18.0.0:Language.Haskell.TH.Quote.quoteExp
              absdir "/"
    • In the quasi-quotation: [absdir|/|]
    |
106 |       [] -> pure (Left [absdir|/|])
    |
```

```
src\Krank\Checkers\Ignore.hs:35:31: error:
    • Exception when trying to run compile-time code:
        src\PyF\Internal\Meta.hs:(212,1)-(347,76): Non-exhaustive patterns in function translateTHtoGHCExt

      Code: template-haskell-2.18.0.0:Language.Haskell.TH.Quote.quoteExp
              fmt
              "Impossible case, update the guard with: {ByteString.unpack command}"
    • In the quasi-quotation:
        [fmt|Impossible case, update the guard with: {ByteString.unpack command}|]
   |
35 |           | otherwise = error [fmt|Impossible case, update the guard with: {ByteString.unpack command}|]
   |
```

</details>

<details>
  <summary>Click to show all 70 excluded packages</summary>

```
airship ==0.9.5,
aura ==3.2.9,
beam-sqlite ==0.5.1.2,
bimap-server ==0.1.0.1,
bindings-uname ==0.1,
blas-hs ==0.1.1.0,
brick ==0.69,
bytestring-mmap ==0.2.2,
check-email ==1.0.2,
climb ==0.3.3,
connection-pool ==0.2.2,
core-program ==0.5.0.4,
core-telemetry ==0.2.3.5,
cryptonite-openssl ==0.7,
curl ==1.3.8,
cursor-brick ==0.1.0.1,
dbus ==1.2.24,
dbus-hslogger ==0.1.0.1,
debian ==4.0.2,
download ==0.3.2.7,
download-curl ==0.1.4,
editor-open ==0.6.0.0,
exit-codes ==1.0.0,
faktory ==1.1.2.2,
fdo-notify ==0.3.1,
filter-logger ==0.6.0.0,
glob-posix ==0.2.0.1,
hdaemonize ==0.5.6,
hfsevents ==0.1.6,
hledger-iadd ==1.3.17,
hsyslog ==5.0.2,
hxt-curl ==9.1.1.1,
katip ==0.8.7.2,
katip-logstash ==0.1.0.2,
katip-wai ==0.1.2.0,
ki ==0.2.0.1,
krank ==0.2.3,
linenoise ==0.3.2,
logging-facade-syslog ==1,
magic ==1.1,
microlens-process ==0.2.0.2,
nvim-hs ==2.2.0.2,
OpenAL ==1.7.0.5,
pager ==0.1.1.0,
path-extra ==0.2.0,
pid1 ==0.1.3.0,
poll ==0.0.0.2,
posix-paths ==0.3.0.0,
posix-pty ==0.2.2,
projectroot ==0.2.0.1,
rainbow ==0.34.2.2,
rainbox ==0.26.0.0,
rawfilepath ==1.0.1,
resolv ==0.1.2.0,
safe-coloured-text-terminfo ==0.0.0.0,
safeio ==0.0.5.0,
sandwich ==0.1.0.9,
shared-memory ==0.2.0.0,
shell-conduit ==5.0.0,
simple-cmd ==0.2.6,
status-notifier-item ==0.3.1.0,
systemd ==2.3.0,
termbox ==0.3.0,
terminfo,
unix,
unix-bytestring ==0.3.7.7,
vty ==5.35.1,
warp-tls-uid ==0.2.0.6,
Xauth ==0.1,
xdg-desktop-entry ==0.1.1.1
```

</details>

### Impact assessment
I have only applied this [patch](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9060) to GHC 9.2. I have left Bifoldable and Bitraversable unchanged.

The build errors all relate to missing Functor instances. Some of them also had Bifoldable and Bitraversable instances but no Foldable and Traversable instances.

Besides that I saw that for `GenericBifunctor` from the `generic-functor` package the docs do _not_ suggest to derive the Functor instance via `GenericBifunctor` but via `GenericFunctor`. If Functor is going to be a superclass of Bifunctor, shouldn't it be advised to derive the required superclass instance via the same newtype? That is the only documentation issue I noticed.

I also saw that `Tree` in `Data.Geometry.QuadTree.Tree` from the `hgeometry` package has Bifoldable1 and Bitraversable1 instances but no Foldable1 and Traversable1 instances.

Perhaps it makes sense to define default methods similar to [`fmapDefault`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Traversable.html#v:fmapDefault) and [`foldMapDefault`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Traversable.html#v:foldMapDefault) in [`Data.Bitraversable`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Bitraversable.html) to derive Bifunctor and Bifoldable in terms of an existing Bitraversable instance?
```
  foldMapDefault = bifoldMap (const mempty)
  traverseDefault = bitraverse pure
```

Here is the full list of build errors:
- `bifunctors` 
  - already fixed in the branch of the not yet released major version 6.
- `dual`
  - `Dual` in `Control.Category.Dual` (no Functor, Foldable, Traversable instance)
- `either-both`
  - `Either'` in `Data.Either.Both` (no Functor, Foldable, Traversable instance)
- `fgl`
  - `Gr` in `Data.Graph.Inductive.Tree` (no Functor instance)
  - `Gr` in `Data.Graph.Inductive.PatriciaTree` (no Functor instance)
- `foundation`
  - `Tuple2` in `Foundation.Tuple` (no Functor instance)
- `generic-functor`
  - `GenericBifunctor` in `Generic.Functor.Internal` (no Functor instance) 
- `pgp-wordlist`
  - `AltList` in `Data.Text.PgpWordlist.Internal.AltList` (no Functor instance)
- `aura`
  - `These` in `Aura.Utils` (no Functor instance)
- `ForestStructures`
  - `SPForest` in `Data.Forest.StructuredPaired` (no Functor, Foldable, Traversable instance)
- `hgeometry-combinatorial`
  - `(:+)` in `Data.Ext` (no Functor, Foldable, Traversable instance)
  -  `Alternating` in `Data.List.Alternating` (no Functor, Foldable, Traversable instance)
  - `Gr` in `Data.PlanarGraph.AdjRep` (no Functor instance)
  - `Vtx` in `Data.PlanarGraph.AdjRep` (no Functor instance)
  - `TreeNode` in `Data.Tree.Util`  (no Functor, Foldable, Traversable instance)
- `hgeometry`
  - `NodeData` in `Data.Geometry.PrioritySearchTree` (no Functor instance)
  - `PrioritySearchTree` in `Data.Geometry.PrioritySearchTree` (no Functor instance)
  - `Box` in `Data.Geometry.Box.Internal` (no Functor, Foldable, Traversable instance)
  - `Tree` in `Data.Geometry.QuadTree.Tree` (no Functor, Foldable, Traversable instance)
  - `Triangle` in `Data.Geometry.Triangle` (no Functor, Foldable, Traversable instance)

**Comment by Bodigrim** [comment]
Thanks @Topsii. Could you plesase share the patches you made for affected packages? For example, fork them, patch and provide `cabal.project` pointing to patched repos. This way someone with a UNIX machine would be able to continue your work and check the rest of the packages. It is also important for CLC to see the patches in full to understand how much effort are required on average to implement this proposal.

That said, the amound of breakage does not seem prohibiting to me, and all affected package can be patched in a backwards-compatible way. 

What's your stance on https://github.com/haskell/core-libraries-committee/issues/91#issuecomment-1256716483, @Icelandjack?

**Comment by Bodigrim** [comment]
Thanks @Topsii, excellent work!

> The impact assessment for Foldable and Traversable uncovered the following build failures.

Are you referring to #93? Let's leave it for another day and stick to the `Bifunctor` proposal at hand for time being.

> I assume the goal is to remove second from Bifunctor at some point in the future?

I don't think so, and the proposal does not mention it. We have not removed `mappend` from `Monoid` or `return` from `Monad`, so I do not see `second` removed from `Bifunctor` in foreseeable future.

----

With regards to `instance Bifunctor (,)` I realised that the proposal does not make things worse than they are right now. [Haddocks](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Bifunctor.html#t:Bifunctor) already require that `first id = id` and `second id = id`, which is not precisely true for tuples. That's pretty unfortunate indeed, but it is pre-existing state of affairs and should be resolved separately of this proposal. 

That said, I don't see any specific obstacles to move the proposal forward. @Icelandjack @treeowl any comments?

**Comment by treeowl** [comment]
As long as we actually do something about the strictness issue.

**Comment by Bodigrim** [comment]
(The issue with `first id` for tuples is the same as with `fmap id` for `NonEmpty` - both are lazy beyond measure. It would be nice to fix it somehow, but that's really a separate issue common for entities from kmettoverse)

**Comment by Bodigrim** [comment]
> I think we should not take this liberty if there is no plan to remove `second`. In my previous comment I gave an example how it can go wrong.

Your example demonstrates that changing the implementation of `second` from `bimap id` to `fmap` is dangerous, because it can cause recursive loops in instances, which were sane previously. Makes sense, thanks. 

@Icelandjack we can't really proceed any further without an active proposer. Are you still interested to pursue this? @Topsii would you like to take over?

**Comment by Icelandjack** [comment]
Yes @Topsii can take over

**Comment by Bodigrim** [comment]
@Topsii are you interested to take over this proposal?

**Comment by Bodigrim** [comment]
Thanks for detailed impact analysis, that's very thorough.

The proper path to remove `second` is:
1. CLC proposal to make `Functor` a super class for `Bifunctor` (we are here).
2. **GHC** proposal to introduce `-Wnoncanonical-bifunctor-instances`, because introduction of new warnings is outside of CLC remit.
3. CLC proposal to remove `second` from `Bifunctor`.

I'd suggest to stick to Step 1 only in this topic, otherwise the discussion will sidetrack pretty quickly. We can return to Step 3 once Step 2 is completed.

At the moment I personally lean against removing `second` from `Bifunctor`: in my opinion potential benefits are not sufficiently pronounced.

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on the proposal to add a `Functor` superclass to `Bifunctor`:
```diff
-class Bifunctor p where
+class (forall a. Functor (p a)) => Bifunctor p where
```

The merge request is available at https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9313/diffs. It does not change the default definition of `second` and does not shrink `{-# MINIMAL #-}` set of definitions. 

The impact analysis can be found at https://github.com/haskell/core-libraries-committee/issues/91#issuecomment-1278299233. The proposer has raised PRs to pretty much all affected packages, adding `instance Functor (T a)` for each `T` such that `instance Bifunctor T` is defined. This is a straightforward and backward-compatible fix. 

@tomjaguarpaw @chessai @emilypi @cgibbard @mixphix 

----

+1 from me. I think this change fits nicely into other recent developments wrt quantified superclasses.

**Comment by tomjaguarpaw** [comment]
-1

----

This proposal is technically a good one. My -1 is not a response to this proposal in particular, but to `Bifunctor` in general, which I don't think we should perpetuate by trying to improve it.

My objection to `Bifunctor` in general is that behaviour that is based on the accident of type parameter ordering seems far too fragile. Besides, where are we going to stop? `Trifunctor`, `Quadrafunctor`? `Functor` is a necessity for several reasons, and remains within the realm of the manageable because it only refers to a single type parameter, and so many useful instances only have a single parameter anyway. I would always rather see `over _Left f` and `over _Right g` (or the equivalent for other data types) than `first f` and `second g`.

**Comment by Ericson2314** [comment]
@tomjaguarpaw I am also confused why `Bifunctor` was added to `base`, but so long as it is there this design seems strictly better? I would advocate both for this change and removing it from `base`, do you see a contradiction between those?

**Comment by emilypi** [comment]
+1

While I do think there are better abstractions, like the `FunctorOf` work that @Icelandjack has been working on, I'd rather not let better be the enemy of good; we should improve the existing bifunctor so that we can unblock the addition of `Profunctor` in `base`, which will *significantly deduplicate many repeated parts of the optics ecosystem*. Later, if there are better paradigms, we can make the switch. This to me seems both minimally painful, and better bang for the buck than the competing ideas. 

With competing ideas, we also have to consider that quite a bit of performance-sensitive code is written with the current `Bifunctor`/`Profunctor` split, particularly in `lens`. Any move towards better abstraction needs to preserve these, or sacrifice performance gains with a significantly better UX/DX. I'm not convinced it will. It seems more like it would just trigger a whole lot of work recalibrating what was already stable, as opposed to the change detailed in this issue, which should have minimal impact @tomjaguarpaw.

**Comment by Icelandjack** [comment]
@emilypi Adding the superclass also gives a path to evolve `Bifunctor` to a `FunctorOf` representation. The goal of `FunctorOf` is to provide a shared functorial interface and eliminating the need for defining a separate class for each point in the hierarchy. Nobody wants a standalone `Trifunctor` class.

**Comment by Icelandjack** [comment]
This is not official, only my personal roadmap. It depends on the ability to be able to define proper type class synonyms, which I think of in terms of class frontends and backends. (I will propose this as it's own feature)

The categorical functor provides an amazingly general interface. The problem is writing instances, we want to continue to use `Bifunctor` for backwards compatibility.

```haskell
-- frontend
instance Bifunctor (,) where
  first :: (a -> a') -> ((a, b) -> (a', b))
  first f (a, b) = (f a, b)
```

rather than exposing natural transformations to the users:

```haskell
-- backend
instance Functor (,) where
  type Source (,) = (->)
  type Target (,) = Nat (->) (->)
  
  fmap :: (a -> a') -> Nat (->) (->) ((,) a) ((,) a')
  fmap f = Nat \(a, b) -> (f a, b)
```

The basic idea is that the frontends don't exist as their own classes. They only elaborate to the backends. If users define their own `Trifunctor` frontends they can interoperate because they program against the same interface. Also any `FunctorOf` interface that is more general than `Trifunctor` is still compatible. I would make `FunctorOf` a MPTC frontend of the categorical `Functor` (eliding methods is a shorthand for when they are the same):

```haskell
type  FunctorOf :: Cat s -> Cat t -> (s -> t) -> Constraint
class frontend FunctorOf src tgt f where
  instance Functor f where
    type Source f = src
    type Target f = tgt
```

This instance gets elaborated into the same `Functor (,)` instance as above:

```haskell
instance FunctorOf (->) (Nat (->) (->)) (,) where
  fmap :: (a -> a') -> Nat (->) (->) ((,) a) ((,) a')
  fmap f = Nat \(a, b) -> (f a, b)
```

Now we can define `Bifunctor` as a frontend. @tomjaguarpaw: In order to construct the natural transformation between `bi a` and `bi a'` we must require that both of them are regular Haskell `Functor`s, namely `FunctorOf (->) (->)`. So in order to elaborate the `Bifunctor bi` frontend to the categorical functor we need some guarantee that `bi a` is a functor for all `a`. This is exactly why the quantified constraint is necessary because otherwise a `Bifunctor` class does not have enough information to be elaborated into a categorical functor.

```haskell
type  Bifunctor :: (Type -> Type -> Type) -> Constraint
class frontend (forall a. Functor (bi a)) => Bifunctor bi where
  first :: (a -> a') -> (bi a b -> bi a' b)
  
  instance FunctorOf (->) (Nat (->) (->)) bi where
    fmap :: (a -> a') -> Nat (->) (->) (bi a) (bi a')
    fmap f = Nat (first f)
```

See the [definition of `Nat`](https://gist.github.com/ekmett/b26363fc0f38777a637d)

```haskell
type Nat :: Cat s -> Cat t -> Cat (s -> t)
data Nat src tgt f g where
  Nat :: (FunctorOf src tgt f, FunctorOf src tgt g) => (forall a. tgt (f a) (g a)) -> Nat src tgt f g
```

This is a longshot but without this functionality I don't see any way to evolve our hierarchies. As an added bonus this separates the *interface* of a class from its machine *representation* so we can derive `Traversable` and every other type class you can think of! (see why this is a [pain](https://ryanglscott.github.io/2018/06/22/quantifiedconstraints-and-the-trouble-with-traversable/)).

The same will be possible for `Generic` and `Generic1` which face the same arbitrariness: why not `Generic2`.. etc.? Instead we introduce [`GenericK` for arbitrary kinds](https://hackage.haskell.org/package/kind-generics) as a backend for `Generic` and `Generic1`. Every existing instance continues to compile as usual.

Then a lot of other classes like [`Filterable`](https://hackage.haskell.org/package/witherable-0.4.2/docs/Witherable.html#t:Filterable) can be elaborated into `FunctorOf (Kleisli Maybe) (->)` with a minor technicality that you need a newtype wrapper so it doesn't overlap with `FunctorOf (->) (->)` instances but this is solved with a newtype that the user never has to worry about.

**Comment by Bodigrim** [comment]
FWIW the proposal makes `Bifunctor` more restrictive than currently, which is a good thing even (and especially) if you do not like the class itself. 

@cgibbard @chreekat @mixphix this is a gentle reminder to vote.

**Comment by chessai** [comment]
+1

I also think it should be removed, but we should improve the quality of what is already present

**Comment by tomjaguarpaw** [comment]
+1

----

OK, I am persuaded to change my vote, if that is permitted. The notion of "bifunctor" is a perfectly coherent one, I just don't particularly like the current implementation (and better implementations seem somewhat elusive in the current type system of Haskell). This proposal actually improves the current implementation (although only a small amount) so perhaps best not to oppose it.

N.B. my feelings about `Bifunctor`, which I believe is an effort at an implementation of a coherent notion, are in contrast with my feelings about `Foldable`, which I believe isn't.

**Comment by mixphix** [comment]
+1, and thank you for the detailed explanation of `FunctorOf` @Icelandjack !

**Comment by Bodigrim** [comment]
This gives us 5 votes in favor out of possible 6, so the proposal is approved. Thanks all!

**Comment by Bodigrim** [comment]
@Topsii one last thing, could you please prepare a migration guide for this breaking change? You can probably use https://github.com/haskell/core-libraries-committee/blob/main/guides/functor-combinator-instances-and-class1s.md as a template.

**Comment by Ericson2314** [comment]
How do you feel about `fmap = bimap id`?

**Comment by Ericson2314** [comment]
@Topsii what confuses me is that today's warnings are about the redundant method is the *downstream* class. Analogously. they would care about something `second = ...` other than `second = fmap`, but they wouldn't care how `fmap` is defined.

**Comment by Ericson2314** [comment]
I think it should be perfect allowable to just define the maximally downstream classes one wishes too, and derive all the upstream class impls, so `fmap = bimap id` or `sconcat = ... mconcat .... should be fine` right?

**Comment by Bodigrim** [comment]
Adding a GHC warning for non-canonical instances is quite a big deal. To implement it `Bifunctor` has to become a magical type class, implemented in `ghc-prim`, not in `base`. 

We are in early days of using `QuantifiedConstraints` for type class hierarchy. I’d wait until this proposal is battle-tested and the destiny of proposals for Bifoldable / Bitraversable / Profunctor is determined, before making a GHC proposal.

For migration guide I’d recommend users to move their definition of `second` to `fmap` and redefine `second = fmap` afterwards.

**Comment by Bodigrim** [comment]
Most of the time `instance Functor` can be derived, no need to implement `fmap` manually.

**Comment by Ericson2314** [comment]
@Topsii I think those warnings on "defined backwards" are just for when the redundant method is used verbatim. So `second = fmap` would fall under that, but `second = bimap id` would not.

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Authors**            |  @Topsii, @Icelandjack |
| **Status**             |  merged  |
| **`base` version**     | `4.18.0.0`   |
| **Merge Request (MR)** | https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9313 |
| **Blocked by**         | nothing |
| **CHANGELOG entry**    | present  |
| **Migration guide**    |  https://github.com/haskell/core-libraries-committee/blob/main/guides/bifunctor-superclass.md |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂

**Comment by hasufell** [comment]
> The build errors all relate to missing Functor instances. Some of them also had Bifoldable and Bitraversable instances but no Foldable and Traversable instances.

It seems our impact assessments assume that patches will be merged. What if they won't, e.g. because it was a conscious decision to omit an instance. 

https://github.com/haskell-foundation/foundation/pull/572

We should be more careful with things that impact alternative preludes.

**Comment by Bodigrim** [comment]
(Vincent banned me long ago, so let me comment here)

> It seems our impact assessments assume that patches will be merged. What if they won't, e.g. because it was a conscious decision to omit an instance.

As a matter of principle: we cannot allow a single maintainer, unwilling to accept simple patches, to block changes, which we otherwise find desirable. `foundation` has long been known as a liability and a laggard to support new GHCs, even despite community's efforts to provide necessary patches promptly. I've seen this drama with GHC 9.0, and 9.2, and 9.4, and by 9.6 it grew too old. People should have migrated away long ago.

On this specific instance: I know that some people still find `instance Foldable (a,)` controversial because of `length` behaviour (although we are almost a decade past this choice, go use Hugs if you wish). But there is no controversy with `instance Functor (a,)` and no reason not to define it.

**Comment by Bodigrim** [comment]
For the sake of any new readers, I should also mention that there've been plenty of time to raise concerns about this proposal. The discussion has started in September 2022, the PR for `foundation` was raised a month later, in October, giving maintainers an explicit opportunity to voice an opinion, and finally the vote has been held in the late November. But if one kept silence all this time, it's unreasonable to get angry about the change half a year later. 

I've seen this pattern before in https://github.com/haskell/core-libraries-committee/issues/3#issuecomment-968286173, and it makes me sad. If you are unhappy with a proposal, please do chime in, have your say here, not just vent on Twitter or gnash teeth silently.

----

Note that `foundation` is not consistent about not providing `Functor` instances for types with two independent parameters: e. g., [`These a b`](https://hackage.haskell.org/package/foundation-0.0.29/docs/Foundation-Primitive.html#t:These) has `instance Functor (These a)`. Judging from this, not having `instance Functor (Tuple a)` looks an inadverent omission, not a general principle or belief.

**Comment by Ericson2314** [comment]
Very much agree with both of those posts.

And for what it is worth `Bifunctor` is exactly the sort of thing I would like to see decoupled from GHC, so we can continue improving our libraries and not breaking any stick-in-the-muds or laggards. The best of both worlds.

**Comment by vincenthz** [comment]
> For the sake of any new readers, I should also mention that there've been plenty of time to raise concerns about this proposal. The discussion has started in September 2022, the PR for `foundation` was raised a month later, in October, giving maintainers an explicit opportunity to voice an opinion, and finally the vote has been held in the late November. But if one kept silence all this time, it's unreasonable to get angry about the change half a year later.

Wow. between a whole *month* and two months to react, how gracious.

The lack of shipping those potential changes in alpha/beta forms to users, means maintainers need to keep on top of the bleeding edge development (and PR if they are lucky) until it's pushed as 'stable' to the unsuspecting users as a new base release.

It's funny how some other languages have understood this balance very well, and enjoy way more successes in the real world ... but I guess I'm the one 'unreasonable'

> Note that `foundation` is not consistent about not providing `Functor` instances for types with two independent parameters: e. g., [`These a b`](https://hackage.haskell.org/package/foundation-0.0.29/docs/Foundation-Primitive.html#t:These) has `instance Functor (These a)`. Judging from this, not having `instance Functor (Tuple a)` looks an inadverent omission, not a general principle or belief.

you're wrong, simple overzealous mirroring of the existing datatype from the 'these' package. Funny how you can 'judge' things without knowledge for your benefits when that suits you.

**Comment by tomjaguarpaw** [comment]
> lack of shipping those potential changes in alpha/beta forms

By the way, I believe this is now available through GHC nightlies. See, for example, https://www.haskell.org/ghcup/guide/#nightlies
